common/src/main/java/io/netty/util/internal/telemetry/EliminationHookQueue.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util.internal.telemetry;
+
+import io.netty.util.internal.QueueFactory;
+import io.netty.util.internal.UnstableApi;
+
+import java.util.AbstractQueue;
+import java.util.Iterator;
+import java.util.Queue;
+
+@UnstableApi
+public class EliminationHookQueue<E> extends AbstractQueue<E> {
+    private final Queue<EleminationHookQueueEntry<E>> queue;
+    private final QueueEliminationStrategy eliminationStrategy;
+
+    public EliminationHookQueue(QueueFactory queueFactory,
+                                QueueEliminationStrategy eliminationStrategy) {
+        queue = queueFactory.newQueue();
+        this.eliminationStrategy = eliminationStrategy;
+    }
+
+    public QueueEliminationStrategy getEliminationStrategy() {
+        return eliminationStrategy;
+    }
+
+    @Override
+    public Iterator<E> iterator() {
+        final Iterator<EleminationHookQueueEntry<E>> iterator = queue.iterator();
+        return new Iterator<E>() {
+            @Override
+            public boolean hasNext() {
+                return iterator.hasNext();
+            }
+
+            @Override
+            public E next() {
+                return iterator.next().getEntry();
+            }
+        };
+    }
+
+    @Override
+    public int size() {
+        return queue.size();
+    }
+
+    @Override
+    public boolean offer(E e) {
+        return queue.offer(new EleminationHookQueueEntry<E>(e));
```

Comment:
This is a reasonable approach to measure latencies per task but it has the downside of one extra object allocation per task submitted to the queue. This may be a bit much for certain high throughput environments. In the past I have seen some folks use an approach where they add a task to the queue and measure latency to execute that task. You can choose when to add the task based on what suits your needs, ranging from re-submitting a task as soon as it is executed to submitting at a routine interval.

Reply:
Hello, Nitesh! Thanks for valuable feedback and good advices, I will improve code according to this.
But I don't fully understand this comment about latency measurement, particularly the last 2 sentences, sorry, could you please explain it? We may have different strategies of tasks execution, but my aim was to work only on latency measurement.
On the low level, other solution that I see here is to declare separate queue and don't wrap runnable instances into objects with insertion time, store those times in a separate queue. This way we will avoid extra object creation (only store Longs in a new queue), but it may require extra synchronization to calculate proper results (add/remove items into 2 queues atomically). However, we may avoid this synchronization and choose speed over accuracy (use the idea that here we may swap insertion times for 2 neighbour tasks, but as this times are really similar we can ignore this fact)

Reply:
I believe what @NiteshKant is another metric people could be interested in ie how much time it takes for a task to be submitted and executed? 
But I think too that this is the kind of stuff that users can be built externally to Netty, given that there is no special need of any of the Netty internals: I believe event  loop telemetry should be able to help user providing metrics that cannot be obtained from inside 
eg 
- how much a ByteBuf write/flush takes before it's getting completed? 
- how much time a read takes from being read from socket to be processed? (no idea if makes sense)
- how much time is spent while performing pure CPU tasks vs I/O tasks on the event loop?

Reply:
I think are two distinct flavor of these metrics:

- Ones that are global; like latencies for all tasks submitted to a queue (what this change is doing)
- Ones that are specific to a type; like what @franz1981 lists

I believe anything global is relatively easier and can be done by users without a need for a change to netty internals. My example was one such way. On the other hand, the metrics for specific types like end-to-end read latency may require additional instrumentation and this is perhaps we can focus on to see whether there are hooks available already for folks to do it or not.

Reply:
@fedorovr 

> But I don't fully understand this comment about latency measurement, particularly the last 2 sentences, sorry, could you please explain it?

I am sorry, should have been clearer.

What this change is trying to do is to determine "time spent in the queue" and it takes the approach of measuring latency for each task between `offer()` and `poll()`. The approach that I mention takes an alternate approach considering that if the queue is empty a task should be executed immediately otherwise measuring "time spent in the queue" for an arbitrary task in the queue should give us a descent insight into queue depths over time. Submitting a "dummy task" to the queue and measuring its latency essentially serves the purpose of the arbitrary task. The last 2 lines in my original message were just proposing different approaches of when to submit this "dummy task"

Reply:
Got it, thanks for explanation. This idea seems to be more elegant and more lightweight. I will look at the code to check possible implementation

Reply:
We currently use approach with "dummy task", periodically send it to worker to measuring latency. All this stuff build on top of `SingleThreadEventExecutor` and ability to get `ThreadProperties` which is contain information about thread state.